fix(node): canonical names in policy generation

naugtur opened the pull request:
Added two independent tests for canonical names being used in policy and fixed policy generation.

Adopted the label field for canonical names. This is a clean solution, the name "label" is unfortunate but it was no less unfortunate before. The meanings it factually had didn't change.

TODO:
 - [x] Adjust some assertions to work with canonical names
 - [x] doublecheck if there's redundancy that can now be removed from the policy generation process.

naugtur committed:
test(node): test canonical names in policy generation e2e

naugtur committed:
fix(node): generate policy with canonical names

boneskull commented on .gitignore:
```diff
@@ -122,6 +122,9 @@ packages/*/types
 
 !**/fixtures/main/node_modules
+
+!packages/node/test/e2e/fixture/deptree/**/node_modules
```
note to self: I think we can exempt any `node_modules` in `packages/node/test/*/fixture` from ignore

boneskull commented on packages/node/src/policy-gen/policy-gen-compartment-map.js:
```diff
@@ -36,6 +36,24 @@ export const loadCompartmentMap = async (
   })
 
+  // In lavamoat the canonical name is what we present to the end user as a resource key in policy. We have no use for Endo's label field, so we replace labels with canonicalNames derived from path.
+  // NOTE: the algorithm creating paths happens to be identical to the one in @lavamoat/aa package. Not that it matters because policies cannot be reused between this and other lavamoat tools.
```
@naugtur

> Not that it matters because policies cannot be reused between this and other lavamoat tools.

I'm not sure I follow; do you mean that a policy created for a node app won't be helpful if you try to bundle that app and run it in the browser?  not that the policy _schema_ is incompatible?

boneskull committed:
chore(core): refactors and improvements for test harnesses

- Modifies the `testPolicyForJSON` macro to support arbitrary assertions.
- `runCLI()` accepts an `ExecutionContext` and logs the command (instead of the `testCLI` macro); returns the command
- Created some types and updated a whole lot of docstrings
- Remove crap about testing inline `Buffer`s. What was I thinking?
- Add `keysOr()` and `isFunction()` utils

boneskull committed:
chore(node): update policy gen tests for canonical names

- fix CLI tests to use `--root` (since `runCLI()` does not execute in any particular CWD)
- `test/unit/policy-gen.spec.js` covered by `test/unit/policy/generation.spec.js` via macro
- update snapshots
- bonus: don't need to run scenario tests serially!

boneskull committed:
chore(node): consolidate getting of canonical name

boneskull committed:
chore(node): disable wacky tseslint rule

naugtur commented on packages/node/src/policy-gen/policy-gen-util.js:
```diff
@@ -60,7 +58,7 @@ export const getPackageName = (compartment, isEntry = false) => {
     return LAVAMOAT_PKG_POLICY_ROOT
   }
-  return compartment.name
+  return compartment.location
```
Should that not be label instead? Do we want to rely on Endo copying over label to location? It's not as simple as a plain copy afair.

boneskull replied:
...yeah, it should be label.  But IIRC this function gets completely blown away downstack.

boneskull committed:
chore(node): fix merge issue

naugtur commented on packages/node/test/unit/scenario.spec.js:
```diff
@@ -47,5 +47,5 @@ for await (const scenario of loadScenarios()) {
     continue
   }
 
-  test.serial(scenario.name ?? '(unknown scenario)', testScenario, scenario)
```
Why do we want that to no longer be serial? Or why did we need it before?

boneskull replied:
It may have been required at some point, but I was probably just mistaken. Anyway, it goes faster now. scenarios go brrrrr

naugtur reviewed (commented):
I'm not allowed to approve a PR I opened, but consider this my conditional approval.
The comment about .location vs .label needs attention before merging.
I didn't review policy-macros in tests

boneskull reviewed (approved):
I approve my changes on behalf of myself
